redis/tests/test_cache.rs

```diff
@@ -0,0 +1,307 @@
+            })
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        assert_eq!(s.miss, 1000);
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_guards_and_different_types() {
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let mut con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+        let _: () = con.set("KEY", "77").await?;
+        let mut v = vec![];
+        for _ in 0..5 {
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: u32 = c.get("KEY").await.unwrap();
+                assert_eq!(x, 77);
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: String = c.get("KEY").await.unwrap();
+                assert_eq!(x, "77");
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: u8 = c.get("KEY").await.unwrap();
+                assert_eq!(x, 77);
+            }));
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        // It should get value once and miss then it should use cache.
+        assert_eq!(s.hit, 14);
+        assert_eq!(s.miss, 1);
+        // 1 GET + 1 PTTL
+        assert_eq!(s.sent_command_count, 2);
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_guards_and_nil_values() {
+    // Testing guard will work even with Nil Value is returned from server
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+        let mut v = vec![];
+        for _ in 0..5 {
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<u32> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<String> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<u8> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        // It should get value once and miss then it should use cache.
+        assert_eq!(s.hit, 14);
```

Comment:
why here the requests hit the cache and in `test_cache_drops_and_guards` they don't?